Add xattr saving and restoration to images

snehring opened the pull request:
## Description of the Pull Request (PR):

This PR adds a way to preserve file xattrs (acls, file capabilities, selinux labels) of files in images and overlays. It does this by searching the root of the image/overlay at creation time and recording the xattrs of all files in an unique to each image/overlay file staged in /warewulf/xattrs of the image/overlay. A new init.d script will read these files and apply the xattrs in them and delete them.  When the runtime overlays are updated by wwclient it will restore any xattrs present in the /warewulf/xattrs directory.

Need to implement the getfattr during image creation and setfattr in runtime updates to native golang implementations.


WIP obviously

## This fixes or addresses the following GitHub issues:

- Fixes #730 (potentially)


## Reviewer  checklist

The reviewer checks the following items before merging the PR.

- [ ] The PR is based on the appropriate branch (typically [main](https://github.com/warewulf/warewulf/tree/main/userdocs))
- [ ] All commits are "Signed off" (e.g., using `git commit --signoff`) in agreement to the [DCO](DCO.txt)
- [ ] The [CHANGELOG](https://github.com/warewulf/warewulf/blob/main/CHANGELOG.md) has been updated, if necessary, and under the correct release heading
- [ ] The [userdocs](https://github.com/warewulf/warewulf/tree/main/userdocs) have been updated, if necessary
- [ ] The submitter is listed in the [contributors file](https://github.com/warewulf/warewulf/blob/main/CONTRIBUTORS.md)
- [ ] The test suite has been updated, if necessary

snehring commented on internal/app/wwclient/root.go:
```diff
@@ -321,7 +321,19 @@ func updateSystem(target string, ipaddr string, port int, wwid string, tag strin
+	// apply saved xattrs
+	_, err = os.Stat(filepath.Join(tempDir, "xattrs"))
+	if err != nil {
+		wwlog.Debug("no xattrs file in overlay")
```
This is just applying the runtime overlay right? Isn't that just a monolithic single overlay once composed and sent to the client?

mslacken replied:
@snehring you are right, `wwclient` just manages the runtime overlay. Still I think it would be better to have `_, err = os.Stat(filepath.Join(tempDir, "xattrs.d","runtime"))` so store the xattrs separately for the overlays, so that in [overlays/wwinit/rootfs/warewulf/init.d/80-xattrs](https://github.com/warewulf/warewulf/pull/2043/files#diff-aac772994e5d842b5e5500cdc044aad03093a4626a1e719eba60570e989c4490) we might also have xattrs in the images.
@anderbubble can you please add this to agenda for the next meeting.

snehring commented on internal/pkg/util/util.go:
```diff
@@ -517,6 +517,12 @@ func BuildFsImage(
 	}
 	wwlog.Debug("Created image directory for %s: %s", name, imagePath)
 
+	err = CreateXattrDump(rootfsPath, filepath.Join(rootfsPath, "xattrs"))
```
This is actually probably the solution right here. Make a directory /warewulf/xattrs/ with a unique file created for each overlay, then have 80-xattrs process all files in that directory.

snehring commented on internal/pkg/util/util.go:
```diff
@@ -622,3 +628,34 @@ func EqualYaml(a interface{}, b interface{}) (bool, error) {
+		mask = "security.capability|system.*"
+	}
+
+	getfattr := exec.Command("getfattr", "-R", "-d", "-m", mask, "-e", "hex", "-P", "-h", ".")
```
I'll look around and see what I can find.

snehring committed:
Add xattr saving and restoration to images

Signed-off-by: Shane Nehring <snehring@iastate.edu>

snehring commented:
rebased to current master and squashed

I'm hoping to find some time soon to add a couple tests in, mostly just tyring to figure out the best way to do it.
